src/http.rs

```diff
@@ -181,18 +185,92 @@ pub enum Auth {
 }
 
 impl Auth {
+    pub fn get_and_strip_basic_auth(url: &str) -> Option<(String, Self)> {
```

Comment:
Given the usage of this, a return type of `(String, Option<Self>)` might work better. You already make the relevant decisions here, so it would avoid needing to do the `match` in the caller. Of course, then you couldn't use `?` directly, at least not without breaking this into two parts.